File: frontend/analyticsfrontend/domain-analysis.js
const axios = require('axios');
require('dotenv').config({ path: './env.local' });

/**
 * Domain History and Age Analysis Service
 * Integrates WHOIS and DNS History APIs from WHOISXMLAPIs
 */
class DomainAnalysisService {
    constructor() {
        this.whoisApiKey = process.env.WHOIS_API_KEY;
        this.whoisBaseUrl = process.env.WHOIS_API_BASE_URL;
        this.whoisHistoryBaseUrl = process.env.WHOIS_HISTORY_API_BASE_URL;
        
        if (!this.whoisApiKey) {
            throw new Error('WHOIS_API_KEY is required in env.local file');
        }
    }

    /**
     * Get comprehensive domain information including age and history
     * @param {string} domain - Domain name to analyze
     * @returns {Object} Domain analysis results
     */
    async analyzeDomain(domain) {
        try {
            console.log(`🔍 Analyzing domain: ${domain}`);
            
            // Get WHOIS information
            const whoisData = await this.getWhoisInfo(domain);
            
            // Get WHOIS history for the domain
            const whoisHistory = await this.getWhoisHistory(domain);
            
            // Log raw API responses for debugging
            console.log('\n📊 RAW WHOIS DATA:');
            console.log(JSON.stringify(whoisData, null, 2));
            
            console.log('\n📊 RAW WHOIS HISTORY DATA:');
            console.log(JSON.stringify(whoisHistory, null, 2));
            
            // Calculate domain age
            const domainAge = this.calculateDomainAge(whoisData.createdDate);
            
            // Compile results
            const analysis = {
                domain,
                whois: whoisData,
                whoisHistory,
                domainAge,
                analysis: {
                    isEstablished: domainAge.years >= 2,
                    isVeteran: domainAge.years >= 5,
                    credibility: this.calculateCredibilityScore(whoisData, whoisHistory, domainAge)
                }
            };
            
            console.log(`✅ Analysis complete for ${domain}`);
            return analysis;
            
        } catch (error) {
            console.error(`❌ Error analyzing domain ${domain}:`, error.message);
            throw error;
        }
    }

    /**
     * Get WHOIS information for a domain
     * @param {string} domain - Domain name
     * @returns {Object} WHOIS data
     */
    async getWhoisInfo(domain) {
        const url = `${this.whoisBaseUrl}/whoisserver/WhoisService`;
        const params = {
            domainName: domain,
            outputFormat: 'JSON',
            _hardRefresh: 1,
            apiKey: this.whoisApiKey
        };

        try {
            const response = await axios.get(url, { params, timeout: 10000 });
            
            // Log raw WHOIS API response
            console.log('\n🔍 RAW WHOIS API RESPONSE:');
            console.log(JSON.stringify(response.data, null, 2));
            
            if (response.data && response.data.WhoisRecord) {
                const whois = response.data.WhoisRecord;
                
                // Extract creation date from multiple possible locations
                let createdDate = whois.createdDate;
                let updatedDate = whois.updatedDate;
                let expiresDate = whois.expiresDate;
                let registrar = whois.registrarName;
                let status = whois.status;
                let nameServers = whois.nameServers?.hostNames || [];
                
                // If main data is missing, try to get from registryData
                if (!createdDate && whois.registryData) {
                    createdDate = whois.registryData.createdDate;
                    updatedDate = whois.registryData.updatedDate;
                    expiresDate = whois.registryData.expiresDate;
                    registrar = whois.registryData.registrarName || registrar;
                    status = whois.registryData.status || status;
                    nameServers = whois.registryData.nameServers?.hostNames || nameServers;
                }
                
                return {
                    createdDate,
                    updatedDate,
                    expiresDate,
                    registrar,
                    status,
                    ips: this.extractIPs(whois),
                    nameServers,
                    registrant: whois.registrant?.organization || 'Unknown',
                    country: whois.registrant?.countryCode || 'Unknown'
                };
            } else {
                throw new Error('Invalid WHOIS response format');
            }
        } catch (error) {
            throw new Error(`WHOIS lookup failed: ${error.message}`);
        }
    }

    /**
     * Get WHOIS history for a domain
     * @param {string} domain - Domain name
     * @returns {Object} WHOIS history data
     */
    async getWhoisHistory(domain) {
        const url = `${this.whoisHistoryBaseUrl}/api/v1`;
        const params = {
            domainName: domain,
            outputFormat: 'JSON',
            apiKey: this.whoisApiKey
        };

        try {
            const response = await axios.get(url, { params, timeout: 10000 });
            
            // Log raw WHOIS History API response
            console.log('\n🔍 RAW WHOIS HISTORY API RESPONSE:');
            console.log(JSON.stringify(response.data, null, 2));
            
            if (response.data && response.data.recordsCount !== undefined) {
                return {
                    totalRecords: response.data.recordsCount,
                    firstSeen: null, // API doesn't provide this
                    lastVisit: null, // API doesn't provide this
                    records: [], // API doesn't provide individual records
                    note: 'API returns only count, not individual records'
                };
            } else {
                return null;
            }
        } catch (error) {
            console.warn(`WHOIS History lookup failed for domain ${domain}: ${error.message}`);
            return null;
        }
    }

    /**
     * Calculate domain age from creation date
     * @param {string} createdDate - ISO date string
     * @returns {Object} Age breakdown
     */
    calculateDomainAge(createdDate) {
        if (!createdDate) {
            return { 
                years: 0, 
                months: 0, 
                days: 0, 
                totalDays: 0, 
                createdDate: null,
                ageDescription: 'Unknown'
            };
        }

        try {
            const created = new Date(createdDate);
            
            // Check if date is valid
            if (isNaN(created.getTime())) {
                return { 
                    years: 0, 
                    months: 0, 
                    days: 0, 
                    totalDays: 0, 
                    createdDate: null,
                    ageDescription: 'Invalid Date'
                };
            }
            
            const now = new Date();
            const diffTime = Math.abs(now - created);
            const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
            
            const years = Math.floor(diffDays / 365);
            const months = Math.floor((diffDays % 365) / 30);
            const days = diffDays % 30;

            return {
                years,
                months,
                days,
                totalDays: diffDays,
                createdDate: created.toISOString(),
                ageDescription: this.getAgeDescription(years, months)
            };
        } catch (error) {
            console.warn(`Error calculating domain age for date: ${createdDate}`, error.message);
            return { 
                years: 0, 
                months: 0, 
                days: 0, 
                totalDays: 0, 
                createdDate: null,
                ageDescription: 'Error'
            };
        }
    }

    /**
     * Get age description based on years and months
     * @param {number} years - Number of years
     * @param {number} months - Number of months
     * @returns {string} Age description
     */
    getAgeDescription(years, months) {
        if (years >= 10) return 'Veteran';
        if (years >= 5) return 'Established';
        if (years >= 2) return 'Mature';
        if (years >= 1) return 'Young';
        if (months >= 6) return 'Recent';
        if (months >= 1) return 'New';
        return 'Very New';
    }

    /**
     * Calculate credibility score based on domain analysis
     * @param {Object} whoisData - WHOIS information
     * @param {Object} whoisHistory - WHOIS history data
     * @param {Object} domainAge - Domain age information
     * @returns {number} Credibility score (0-100)
     */
    calculateCredibilityScore(whoisData, whoisHistory, domainAge) {
        let score = 0;
        
        // Domain age score (40 points max)
        if (domainAge.years >= 5) score += 40;
        else if (domainAge.years >= 2) score += 30;
        else if (domainAge.years >= 1) score += 20;
        else if (domainAge.months >= 6) score += 10;
        
        // Registration status score (20 points max)
        if (whoisData.status && !whoisData.status.includes('expired')) {
            score += 20;
        }
        
        // Name servers score (15 points max)
        if (whoisData.nameServers && whoisData.nameServers.length >= 2) {
            score += 15;
        }
        
        // WHOIS history score (15 points max)
        if (whoisHistory && whoisHistory.totalRecords > 0) {
            score += Math.min(15, Math.floor(whoisHistory.totalRecords / 10));
        }
        
        // Registrar reputation score (10 points max)
        if (whoisData.registrar && !whoisData.registrar.includes('Unknown')) {
            score += 10;
        }
        
        return Math.min(100, score);
    }

    /**
     * Extract IP addresses from WHOIS data
     * @param {Object} whois - WHOIS record
     * @returns {Array} Array of IP addresses
     */
    extractIPs(whois) {
        const ips = [];
        
        // Check nameServers.ips array
        if (whois.nameServers && whois.nameServers.ips) {
            ips.push(...whois.nameServers.ips);
        }
        
        // Check registryData nameServers.ips array
        if (whois.registryData && whois.registryData.nameServers && whois.registryData.nameServers.ips) {
            ips.push(...whois.registryData.nameServers.ips);
        }
        
        return ips;
    }
}

module.exports = DomainAnalysisService;

File: backend/src/services/domain_analysis.py
"""
Domain History and Age Analysis Service
Integrates WHOIS and DNS History APIs from WHOISXMLAPIs
"""

import logging
import math
from datetime import datetime, timezone
from typing import Dict, Any, Optional, List

import requests
from ..core.config import get_api_config

logging.basicConfig(level=logging.INFO, format="%(message)s")
log = logging.getLogger("domain-analysis")


class DomainAnalysisService:
    def __init__(self) -> None:
        self.api_config = get_api_config()
        self.whois_api_key = self.api_config.WHOIS_API_KEY
        self.whois_base_url = self.api_config.WHOIS_API_BASE_URL
        self.whois_history_base_url = self.api_config.WHOIS_HISTORY_API_BASE_URL

        if not self.whois_api_key:
            raise RuntimeError("WHOIS_API_KEY is required in environment configuration")

    # ------------------------------------------------------------------ #
    async def analyze_domain(self, domain: str) -> Dict[str, Any]:
        try:
            log.info("🔍 Analyzing domain: %s", domain)

            whois_data = await self.get_whois_info(domain)
            whois_history = await self.get_whois_history(domain)

            log.info("\n📊 RAW WHOIS DATA:")
            log.info("%s", _pretty(whois_data))

            log.info("\n📊 RAW WHOIS HISTORY DATA:")
            log.info("%s", _pretty(whois_history))

            domain_age = self.calculate_domain_age(whois_data["createdDate"])

            analysis = {
                "domain": domain,
                "whois": whois_data,
                "whoisHistory": whois_history,
                "domainAge": domain_age,
                "analysis": {
                    "isEstablished": domain_age["years"] >= 2,
                    "isVeteran": domain_age["years"] >= 5,
                    "credibility": self.calculate_credibility_score(
                        whois_data, whois_history, domain_age
                    ),
                },
            }

            log.info("✅ Analysis complete for %s", domain)
            return analysis

        except Exception as e:
            log.error("❌ Error analyzing domain %s: %s", domain, e)
            raise

    # ------------------------------------------------------------------ #
    async def get_whois_info(self, domain: str) -> Dict[str, Any]:
        url = f"{self.whois_base_url}/whoisserver/WhoisService"
        params = {
            "domainName": domain,
            "outputFormat": "JSON",
            "_hardRefresh": 1,
            "apiKey": self.whois_api_key,
        }

        try:
            resp = requests.get(url, params=params, timeout=10)
            resp.raise_for_status()
            data = resp.json()

            log.info("\n🔍 RAW WHOIS API RESPONSE:")
            log.info("%s", _pretty(data))

            whois = data.get("WhoisRecord", {})
            if not whois:
                raise ValueError("Invalid WHOIS response format")

            created_date = whois.get("createdDate")
            updated_date = whois.get("updatedDate")
            expires_date = whois.get("expiresDate")
            registrar = whois.get("registrarName")
            status = whois.get("status")
            name_servers = (whois.get("nameServers") or {}).get("hostNames", [])

            # Fallback to registryData
            registry = whois.get("registryData", {})
            if not created_date and registry:
                created_date = registry.get("createdDate")
                updated_date = registry.get("updatedDate")
                expires_date = registry.get("expiresDate")
                registrar = registry.get("registrarName") or registrar
                status = registry.get("status") or status
                name_servers = (
                    (registry.get("nameServers") or {}).get("hostNames") or name_servers
                )

            return {
                "createdDate": created_date,
                "updatedDate": updated_date,
                "expiresDate": expires_date,
                "registrar": registrar,
                "status": status,
                "ips": self.extract_ips(whois),
                "nameServers": name_servers,
                "registrant": (whois.get("registrant") or {}).get("organization") or "Unknown",
                "country": (whois.get("registrant") or {}).get("countryCode") or "Unknown",
            }

        except Exception as e:
            raise RuntimeError(f"WHOIS lookup failed: {e}") from e

    # ------------------------------------------------------------------ #
    async def get_whois_history(self, domain: str) -> Optional[Dict[str, Any]]:
        url = f"{self.whois_history_base_url}/api/v1"
        params = {"domainName": domain, "outputFormat": "JSON", "apiKey": self.whois_api_key}

        try:
            resp = requests.get(url, params=params, timeout=10)
            resp.raise_for_status()
            data = resp.json()

            log.info("\n🔍 RAW WHOIS HISTORY API RESPONSE:")
            log.info("%s", _pretty(data))

            if data.get("recordsCount") is not None:
                return {
                    "totalRecords": data["recordsCount"],
                    "firstSeen": None,
                    "lastVisit": None,
                    "records": [],
                    "note": "API returns only count, not individual records",
                }
            return None

        except Exception as e:
            log.warning("WHOIS History lookup failed for domain %s: %s", domain, e)
            return None

    # ------------------------------------------------------------------ #
    def calculate_domain_age(self, created_date: Optional[str]) -> Dict[str, Any]:
        if not created_date:
            return {
                "years": 0,
                "months": 0,
                "days": 0,
                "totalDays": 0,
                "createdDate": None,
                "ageDescription": "Unknown",
            }

        try:
            created = datetime.fromisoformat(created_date.replace("Z", "+00:00"))
        except Exception:
            return {
                "years": 0,
                "months": 0,
                "days": 0,
                "totalDays": 0,
                "createdDate": None,
                "ageDescription": "Invalid Date",
            }

        delta = datetime.now(timezone.utc) - created
        total_days = delta.days
        years, rem = divmod(total_days, 365)
        months = math.floor(rem / 30)
        days = rem % 30

        return {
            "years": years,
            "months": months,
            "days": days,
            "totalDays": total_days,
            "createdDate": created.isoformat(),
            "ageDescription": self.get_age_description(years, months),
        }

    # ------------------------------------------------------------------ #
    def get_age_description(self, years: int, months: int) -> str:
        if years >= 10:
            return "Veteran"
        if years >= 5:
            return "Established"
        if years >= 2:
            return "Mature"
        if years >= 1:
            return "Young"
        if months >= 6:
            return "Recent"
        if months >= 1:
            return "New"
        return "Very New"

    # ------------------------------------------------------------------ #
    def calculate_credibility_score(
        self, whois_data: Dict[str, Any], whois_history: Dict[str, Any], domain_age: Dict[str, Any]
    ) -> int:
        score = 0

        # Domain age
        if domain_age["years"] >= 5:
            score += 40
        elif domain_age["years"] >= 2:
            score += 30
        elif domain_age["years"] >= 1:
            score += 20
        elif domain_age["months"] >= 6:
            score += 10

        # Registration status
        status = whois_data.get("status") or ""
        if status and "expired" not in str(status).lower():
            score += 20

        # Name servers
        if len(whois_data.get("nameServers") or []) >= 2:
            score += 15

        # WHOIS history
        total_records = whois_history.get("totalRecords", 0) if whois_history else 0
        score += min(15, total_records // 10)

        # Registrar reputation
        registrar = whois_data.get("registrar") or ""
        if registrar and "Unknown" not in registrar:
            score += 10

        return min(100, score)

    # ------------------------------------------------------------------ #
    def extract_ips(self, whois: Dict[str, Any]) -> List[str]:
        ips = []

        # nameServers.ips
        ns = whois.get("nameServers") or {}
        ips.extend(ns.get("ips") or [])

        registry = whois.get("registryData", {})
        reg_ns = registry.get("nameServers") or {}
        ips.extend(reg_ns.get("ips") or [])

        return ips


# ---------------------------------------------------------------------- #
# Helper
def _pretty(obj: Any) -> str:
    import json

    return json.dumps(obj, indent=2, ensure_ascii=False)